pkg/controller/security/manager.go

```diff
@@ -0,0 +1,254 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package security
+
+import (
+	"sync"
+	"time"
+
+	"container/heap"
+
+	istiosecurity "istio.io/istio/pkg/security"
+
+	"kmesh.net/kmesh/pkg/constants"
+	"kmesh.net/kmesh/pkg/logger"
+)
+
+type certExp struct {
+	Identity string
+	exp      time.Time
+	index    int
+}
+
+type certCache struct {
+	cert   *istiosecurity.SecretItem
+	refCnt int32
+}
+
+type certsCache struct {
+	certs map[string]*certCache
+	mu    sync.Mutex
+}
+
+type SecretManager struct {
+	caClient *CaClient
+
+	// configOptions includes all configurable params for the cache.
+	configOptions *istiosecurity.Options
+
+	// storing certificates
+	certsCache *certsCache
+
+	// Prioritize certificates based on exp
+	certsQueue *certificateQueue
+}
+
+var log = logger.NewLoggerField("security")
+
+type certificateQueue struct {
+	certs []*certExp
+	mu    sync.Mutex
+}
+
+func newCertCache() *certsCache {
+	return &certsCache{
+		certs: make(map[string]*certCache),
+		mu:    sync.Mutex{},
+	}
+}
+
+func newPriorityQueue() *certificateQueue {
+	return &certificateQueue{
+		certs: make([]*certExp, 0),
+		mu:    sync.Mutex{},
+	}
+}
+
+func (pq *certificateQueue) Push(x interface{}) {
+	n := len(pq.certs)
+	item := x.(*certExp)
+	item.index = n
+	pq.certs = append(pq.certs, item)
+}
+
+func (pq *certificateQueue) Pop() interface{} {
+	old := pq.certs
+	n := len(old)
+	x := old[n-1]
+	old[n-1] = nil // avoid memory leak
+	x.index = -1
+	pq.certs = old[0 : n-1]
+	return x
+}
+
+func (pq *certificateQueue) Len() int {
+	return len(pq.certs)
+}
+
+func (pq *certificateQueue) Less(i, j int) bool {
+	return pq.certs[i].exp.Before(pq.certs[j].exp)
+}
+
+func (pq *certificateQueue) Swap(i, j int) {
+	pq.certs[i], pq.certs[j] = pq.certs[j], pq.certs[i]
+}
+
+func (pq *certificateQueue) addItem(certExp *certExp) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	heap.Push(pq, certExp)
+}
+
+func (pq *certificateQueue) delete(Identity string) *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	for i := 0; i < len(pq.certs); i++ {
+		if pq.certs[i].Identity == Identity {
+			return heap.Remove(pq, i).(*certExp)
+		}
+	}
+	return nil
+}
+
+// Find the top item and pop if it is about to expire
+func (pq *certificateQueue) lookandPop() *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	top := pq.certs[0]
+	if time.Until(top.exp.Add(-10*time.Minute)) <= 0 {
+		return heap.Pop(pq).(*certExp)
+	}
+	return nil
+}
+
+func (s *SecretManager) lookupandStoreCert(Identity string, newCert *istiosecurity.SecretItem) {
+	s.certsCache.mu.Lock()
+	defer s.certsCache.mu.Unlock()
+	// Check if the key exists in the map
+	// If refCnt == 0, then this certificate is about to be deleted, so do not perform a refresh.
+	certCache := s.certsCache.certs[Identity]
+	if certCache != nil && certCache.refCnt != 0 {
```

Comment:
can we check this before we fetchCerts?

Reply:
sure，we can check it before fetchCert, but the check here cannot be omitted.